tokio-trace-core/src/field.rs

```diff
@@ -187,6 +217,51 @@ where
     DisplayValue(t)
 }
 
+// ===== impl AsKey =====
+
+impl<'f> AsKey for Key<'f> {
+    #[inline]
+    fn as_key<'a>(&self, metadata: &'a Meta<'a>) -> Option<Key<'a>> {
+        if metadata == self.metadata {
```

Comment:
This is what makes this branch _slower_ than master --- but, it's necessary to prevent a key from indexing into a span it doesn't actually correspond to. I was able to make this a little less bad by optimizing span comparison a bit.

Reply:
This logic should not be necessary and it also isn't fool proof. Given that file, mod, and line are optional, two separate call sites could have those fields set to None and have the same target / name. This would result in the metadata matching... (though, two separate call sites should not have the same target...).

Anyway, call sites should be `&'static`, so you can use the pointer to differentiate between two call sites.

Reply:
> Anyway, call sites should be &'static, so you can use the pointer to differentiate between two call sites.

That was my original plan. However, this doesn't work for events, where the metadata is not guaranteed to be `'static`  For example, the `log`-compat crate generates metadata for log records with arbitrary lifetimes. I think in most cases, the log metadata that generates that metadata should also be `'static`, but I wasn't sure if using pointer equality here would be trustworthy or not...

That said, `ptr::eq` was certainly the _fastest_ option in my benchmarks. 

In the

Reply:
Can you link to what specifically didn't work?

Reply:
it works in all the tests I've written. I was just concerned that if events are created with dynamically generated metadata, we wouldn't be able to rely on the metadata having the same address. But I'm not sure if that's a likely scenario, so I think this ought to be pointer equality.

Reply:
@carllerche yes, that was why I originally decided not to use reference equality. However, if we only implement PartialEq for `&'static` meta, then we cannot use keys to index the fields of an `Event` (which is not _guaranteed_ to be `'static`, a requirement of the `log` adapter crate). 

In practice, I think a majority (all?) log records which we convert into `Event`s will be `'static`, as they are typically produced by the `log` crate's macros, but the `log` crate doesn't _guarantee_ them to be `'static`.